tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,252 @@
+        /// and we need to subtract some locals that exist in the localVarSig but they do not belongs to the kickoff method
+        /// For now, we're capturing all locals that are hoisted (except known generated locals),
+        /// and we're capturing all the locals that appear in localVarSig with ` LogLocal `
+        /// </summary>
+        /// <typeparam name="TTarget">Instance type</typeparam>
+        /// <param name="instance">Instance object</param>
+        /// <param name="asyncKickoffInfo">The async kickoff info</param>
+        /// <returns>List of locals</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldInfoNameSanitized[] GetHoistedLocalsFromStateMachine<TTarget>(TTarget instance, AsyncKickoffMethodInfo asyncKickoffInfo)
+        {
+            var foundFields = new List<FieldInfoNameSanitized>();
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+            var kickoffParameters = asyncKickoffInfo.KickoffMethod.GetParameters();
+            for (var i = 0; i < allFields.Length; i++)
+            {
+                var field = allFields[i];
+                if (IsStateMachineField(field))
+                {
+                    continue;
+                }
+
+                bool foundInArguments = false;
+                for (int j = 0; j < kickoffParameters.Length; j++)
+                {
+                    if (kickoffParameters[j].Name == field.Name)
+                    {
+                        foundInArguments = true;
+                        break;
+                    }
+                }
+
+                if (foundInArguments == false)
+                {
+                    var indexOfStateMachineSuffix = field.Name.IndexOf(StateMachineNameSuffix, StringComparison.Ordinal);
+                    if (indexOfStateMachineSuffix > 0)
+                    {
+                        foundFields.Add(new FieldInfoNameSanitized(field, SanitizeAsyncHoistedLocalName(field.Name, indexOfStateMachineSuffix)));
+                    }
+                    else
+                    {
+                        foundFields.Add(new FieldInfoNameSanitized(field, field.Name));
+                    }
+                }
+            }
+
+            return foundFields.ToArray();
+        }
+
+        private static string SanitizeAsyncHoistedLocalName(string fieldName, int indexOfStateMachineSuffix)
+        {
+#if NETCOREAPP3_0_OR_GREATER
+            return fieldName.AsSpan()[1..indexOfStateMachineSuffix].ToString();
+#else
+            return fieldName.Substring(1, indexOfStateMachineSuffix - 1);
+#endif
+        }
+
+        private static bool IsStateMachineField(FieldInfo field)
+        {
+            if (field.Name.StartsWith(StateMachineFieldsNamePrefix))
+            {
+                Log.Information("Ignoring local (state machine field): " + field.FieldType.Name + " " + field.DeclaringType?.Name + " " + field.Name);
```

Comment:
Should this be `Log.Debug` instead? Isn't there a risk that it will flood our logs?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26